File: java/src/main/java/cn/zhangls/decoration/linear/StartOffsetItemDecoration.java
package cn.zhangls.decoration.linear;

import android.graphics.Canvas;
import android.graphics.Rect;
import android.graphics.drawable.Drawable;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

/**
 * Adds an offset to the start of a RecyclerView using a LinearLayoutManager or
 * its subclass.
 * <p>
 * If the RecyclerView.LayoutManager is oriented vertically, the offset will be
 * added to the top of the RecyclerView. If the LayoutManager is oriented
 * horizontally, the offset will be added to the left of the RecyclerView.
 * <p>
 * 支持设置 margin 属性
 *
 * @author zhangls
 */
public class StartOffsetItemDecoration extends RecyclerView.ItemDecoration {

  private Drawable mOffsetDrawable;
  private int mOrientation;
  /**
   * 边距
   */
  private int mMargin;

  /**
   * Constructor that takes in a {@link Drawable} to be drawn at the start of
   * the RecyclerView.
   *
   * @param offsetDrawable The {@code Drawable} to be added to the start of
   *                       the RecyclerView
   */
  public StartOffsetItemDecoration(Drawable offsetDrawable) {
    mOffsetDrawable = offsetDrawable;
  }

  /**
   * Constructor that takes in a {@link Drawable} to be drawn at the start of
   * the RecyclerView.
   *
   * @param offsetDrawable The {@code Drawable} to be added to the start of
   *                       the RecyclerView
   * @param margin         边距
   */
  public StartOffsetItemDecoration(Drawable offsetDrawable, int margin) {
    mOffsetDrawable = offsetDrawable;
    mMargin = margin;
  }

  /**
   * Determines the size and location of the offset to be added to the start
   * of the RecyclerView.
   *
   * @param outRect The {@link Rect} of offsets to be added around the child view
   * @param view    The child view to be decorated with an offset
   * @param parent  The RecyclerView onto which dividers are being added
   * @param state   The current RecyclerView.State of the RecyclerView
   */
  @Override
  public void getItemOffsets(@NonNull Rect outRect, @NonNull View view,
                             @NonNull RecyclerView parent,
                             @NonNull RecyclerView.State state) {
    super.getItemOffsets(outRect, view, parent, state);
    if (parent.getChildAdapterPosition(view) > 0) {
      return;
    }

    if (parent.getLayoutManager() instanceof LinearLayoutManager) {
      mOrientation = ((LinearLayoutManager) parent.getLayoutManager()).getOrientation();
      if (mOrientation == LinearLayoutManager.HORIZONTAL) {
        if (mOffsetDrawable != null) {
          outRect.left = mOffsetDrawable.getIntrinsicWidth();
        }
      } else if (mOrientation == LinearLayoutManager.VERTICAL) {
        if (mOffsetDrawable != null) {
          outRect.top = mOffsetDrawable.getIntrinsicHeight();
        }
      }
    }
  }

  /**
   * Draws horizontal or vertical offset onto the start of the parent
   * RecyclerView.
   *
   * @param c      The {@link Canvas} onto which an offset will be drawn
   * @param parent The RecyclerView onto which an offset is being added
   * @param state  The current RecyclerView.State of the RecyclerView
   */
  @Override
  public void onDraw(@NonNull Canvas c, @NonNull RecyclerView parent,
                     @NonNull RecyclerView.State state) {
    super.onDraw(c, parent, state);
    if (mOffsetDrawable == null) {
      return;
    }

    if (mOrientation == LinearLayoutManager.HORIZONTAL) {
      drawOffsetHorizontal(c, parent);
    } else if (mOrientation == LinearLayoutManager.VERTICAL) {
      drawOffsetVertical(c, parent);
    }
  }


  private void drawOffsetHorizontal(Canvas canvas, RecyclerView parent) {
    int parentTop = parent.getPaddingTop() + mMargin;
    int parentBottom = parent.getHeight() - parent.getPaddingBottom() - mMargin;
    int parentLeft = parent.getPaddingLeft();
    int offsetDrawableRight = parentLeft + mOffsetDrawable.getIntrinsicWidth();

    mOffsetDrawable.setBounds(parentLeft, parentTop, offsetDrawableRight, parentBottom);
    mOffsetDrawable.draw(canvas);
  }

  private void drawOffsetVertical(Canvas canvas, RecyclerView parent) {
    int parentLeft = parent.getPaddingLeft() + mMargin;
    int parentRight = parent.getWidth() - parent.getPaddingRight() - mMargin;
    int parentTop = parent.getPaddingTop();
    int offsetDrawableBottom = parentTop + mOffsetDrawable.getIntrinsicHeight();

    mOffsetDrawable.setBounds(parentLeft, parentTop, parentRight, offsetDrawableBottom);
    mOffsetDrawable.draw(canvas);
  }
}

File: kotlin/src/main/java/com/zhangls/decoration/linear/StartOffsetItemDecoration.kt
package com.zhangls.decoration.linear

import android.graphics.Canvas
import android.graphics.Rect
import android.graphics.drawable.Drawable
import android.view.View
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView

/**
 * Adds an offset to the start of a RecyclerView using a LinearLayoutManager or
 * its subclass.
 *
 *
 * If the RecyclerView.LayoutManager is oriented vertically, the offset will be
 * added to the top of the RecyclerView. If the LayoutManager is oriented
 * horizontally, the offset will be added to the left of the RecyclerView.
 *
 *
 * 支持设置 margin 属性
 *
 * @author zhangls
 */
class StartOffsetItemDecoration(private val divider: Drawable, private val margin: Int = 0) : RecyclerView.ItemDecoration() {
    private var orientation: Int = LinearLayoutManager.HORIZONTAL

    /**
     * Determines the size and location of the offset to be added to the start
     * of the RecyclerView.
     *
     * @param outRect The [Rect] of offsets to be added around the child view
     * @param view    The child view to be decorated with an offset
     * @param parent  The RecyclerView onto which dividers are being added
     * @param state   The current RecyclerView.State of the RecyclerView
     */
    override fun getItemOffsets(outRect: Rect, view: View, parent: RecyclerView, state: RecyclerView.State) {
        super.getItemOffsets(outRect, view, parent, state)
        if (parent.getChildAdapterPosition(view) > 0) {
            return
        }

        if (parent.layoutManager is LinearLayoutManager) {
            val linearLayoutManager = parent.layoutManager as LinearLayoutManager
            orientation = linearLayoutManager.orientation
        } else {
            throw UnsupportedOperationException("StartOffsetItemDecoration 只适用于LinearLayoutManager")
        }

        if (orientation == LinearLayoutManager.HORIZONTAL) {
            outRect.left = divider.intrinsicWidth
        } else if (orientation == LinearLayoutManager.VERTICAL) {
            outRect.top = divider.intrinsicHeight
        }
    }

    /**
     * Draws horizontal or vertical offset onto the start of the parent
     * RecyclerView.
     *
     * @param c      The [Canvas] onto which an offset will be drawn
     * @param parent The RecyclerView onto which an offset is being added
     * @param state  The current RecyclerView.State of the RecyclerView
     */
    override fun onDraw(c: Canvas, parent: RecyclerView, state: RecyclerView.State) {
        if (orientation == LinearLayoutManager.HORIZONTAL) {
            drawOffsetHorizontal(c, parent)
        } else if (orientation == LinearLayoutManager.VERTICAL) {
            drawOffsetVertical(c, parent)
        }
    }


    private fun drawOffsetHorizontal(canvas: Canvas, parent: RecyclerView) {
        val parentTop = parent.paddingTop + margin
        val parentBottom = parent.height - parent.paddingBottom - margin
        val parentLeft = parent.paddingLeft
        val offsetDrawableRight = parentLeft + divider.intrinsicWidth

        divider.setBounds(parentLeft, parentTop, offsetDrawableRight, parentBottom)
        divider.draw(canvas)
    }

    private fun drawOffsetVertical(canvas: Canvas, parent: RecyclerView) {
        val parentLeft = parent.paddingLeft + margin
        val parentRight = parent.width - parent.paddingRight - margin
        val parentTop = parent.paddingTop
        val offsetDrawableBottom = parentTop + divider.intrinsicHeight

        divider.setBounds(parentLeft, parentTop, parentRight, offsetDrawableBottom)
        divider.draw(canvas)
    }
}
